### PR TITLE
usagestats: redact info from URLs, send to BigQuery

### PR DESCRIPTION
**Pull request by attfarhan**
This PR re-adds URLs to event payloads sent to BigQuery, with redaction for URL paths and query parameters not explicitly specified for marketing usage. This is being added to ensure we can implement a marketing attribution solution and determine the effectiveness of our campaigns.

URLs were completely removed in the past to make sure we didn't leak any private info to our analytics systems from Sourcegraph Cloud. With redaction in place for everything except marketing UTMs / query params, we can send URLs without leaking any private data.

## Test plan

Added unit tests for the new function. Tested locally that events are properly sent to a mirror of the BQ data warehouse.

### CODE REVIEW

File: internal/usagestats/event_handlers.go [hunk 1]
```diff
@@ -221,3 +229,38 @@ func serializeLocalEvents(events []Event) ([]*database.Event, error) {
...
+// may contain sensitive info on Sourcegraph Cloud. We replace all paths,
+// and only maintain query parameters in a specified allowlist,
+// which are known to be essential for marketing analytics on Sourcegraph Cloud.
+func redactSensitiveInfoFromCloudURL(rawURL string) (string, error) {
+	parsedURL, err := url.Parse(rawURL)
+	if err != nil {
+		return "", err
+	}
+
+	parsedURL.RawPath = "/redacted"
```
Comment (dadlerj) [inline internal/usagestats/event_handlers.go:243]: Would it be acceptable to edit this to:

```
if parsedURL.Host == "https://sourcegraph.com" { 
	parsedURL.RawPath = "/redacted"
	parsedURL.Path = "/redacted"
}
```

Or whatever the host here would be to make this effective?

We could also allowlist the about, learn, docs, etc. sites if that creates risk. 

I totally understand why this is written the way it is, but the rules can be different for those static sites—no private information is stored in _paths_.

**Review by dadlerj — COMMENTED** [review]
Thank you! Left one question for you... Let me know your perspective.

**Review by eseliger — APPROVED** [review]

File: internal/usagestats/event_handlers.go [hunk 2]
```diff
@@ -221,3 +229,38 @@ func serializeLocalEvents(events []Event) ([]*database.Event, error) {
...
+	parsedURL.Path = "/redacted"
+
+	marketingQueryParameters := map[string]string{
+		"utm_source":   "",
+		"utm_campaign": "",
+		"utm_medium":   "",
+		"utm_term":     "",
+		"utm_content":  "",
+	}
```
Comment (unknwon) [inline internal/usagestats/event_handlers.go:252]: This is more idiomatic Go for the use case here:

```suggestion
	marketingQueryParameters := map[string]struct{}{
		"utm_source":   {},
		"utm_campaign": {},
		"utm_medium":   {},
		"utm_term":     {},
		"utm_content":  {},
	}
```

File: internal/usagestats/event_handlers_test.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			have, err := redactSensitiveInfoFromCloudURL(c.url)
+			if err != nil {
+				t.Fatal("Error in redactSensitiveInfoFromCloudURL")
+			}
+			if c.want != have {
+				t.Fatalf("Failed to redact info from Cloud URL, got %s, want %s", have, c.want)
+			}
```
Comment (unknwon) [inline internal/usagestats/event_handlers_test.go:36]: nit: 
```suggestion
			require.NoError(t, err)
			assert.Equal(t, c.want, have)
```

Examples: https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+assert.Equal&patternType=literal

**Review by unknwon — APPROVED** [review]
Left few comments, but nothing blocking.

File: internal/usagestats/event_handlers.go (hunk 1 above)
Comment (attfarhan) [reply internal/usagestats/event_handlers.go:243]: Yeah good point, forgot that we're going to start sending events from other domains now. I'll make it return early for any URL that has a host that isn't https://sourcegraph.com

**Pull request merged**